File: 두 포인터/2470번 두 용액.cpp
#include<iostream>
#include<algorithm>

#define MAX 100000
#define ll long long

using namespace std;

int n;
ll arr[MAX];
pair<int, int> ans;

void input()
{
	cin >> n;
	for (int i = 0; i < n; ++i)
		cin >> arr[i];

	sort(arr, arr + n);
}

void solve()
{
	ll close = 2000000000;
	int left = 0, right = n - 1;

	while (left != right) {
		ll sum = arr[left] + arr[right];
		if (close > abs(sum)) {
			close = abs(sum);
			ans.first = arr[left];
			ans.second = arr[right];
		}

		if (sum >= 0) --right;
		else ++left;
	}

	cout << ans.first << ' ' << ans.second;
}

void init()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
}

int main()
{
	init();
	input();
	solve();

	return 0;
}

File: 이분 탐색/2470번 두 용액.py
import sys

# 전처리 부분
input = sys.stdin.readline
INF = sys.maxsize

# 알고리즘 부분
def binary_search():
    tmp = INF
    left, right = 0, n - 1

    while left < right:
        sum = arr[left] + arr[right]

        if sum == 0:
            return arr[left], arr[right]

        if tmp > abs(sum):
            tmp = abs(sum)
            result = [arr[left], arr[right]]
        
        if sum < 0:
            left += 1
        else:
            right -= 1

    return result

# 메인 코드 부분
n = int(input())
arr = sorted(list(map(int, input().split())))

ans = binary_search()
print(ans[0], ans[1])